File: bloomstack_core/hook_events/packing_slip.py
import frappe
from frappe.model.mapper import get_mapped_doc


def create_stock_entry(packing_slip_doc, method, target_doc = None):
	def set_missing_values(source, target):
		target.purpose = "Material Transfer"
		target.company = frappe.db.get_value("Delivery Note", source.delivery_note, "company")
		packing_warehouse = frappe.db.get_single_value("Delivery Settings", "packing_warehouse")

		for item in target.items:
			if not item.t_warehouse:
				item.t_warehouse = packing_warehouse

	doc = get_mapped_doc("Packing Slip", packing_slip_doc.name, {
		"Packing Slip": {
			"doctype": "Stock Entry",
			"validation": {
				"docstatus": ["=", 1]
			}
		},
		"Packing Slip Item": {
			"doctype": "Stock Entry Detail",
			"field_map": {
				"serial_no": "serial_no",
				"batch_no": "batch_no",
				"source_warehouse": "s_warehouse",
				"target_warehouse": "t_warehouse"
			},
		}
	}, target_doc, set_missing_values)
	doc.save()
	doc.submit()

File: bloomstack_core/public/js/packing_slip.js
frappe.ui.form.on('Packing Slip', {
    delivery_note: function(frm) {
        frappe.db.get_value('Delivery Note', {name: frm.doc.delivery_note}, 'company', (r) => {
            frm.set_value("company", r.company)
        });

        erpnext.queries.setup_queries(frm, "Warehouse", function() {
            return erpnext.queries.warehouse(frm.doc);
		});
    },
    source_warehouse: function(frm) {
		if(frm.doc.source_warehouse) {
			$.each(frm.doc.items || [], function(i, item) {
				frappe.model.set_value(frm.doctype + " Item", item.name, "source_warehouse", frm.doc.source_warehouse);
			});
		}
    },
    target_warehouse: function(frm) {
        if(frm.doc.target_warehouse) {
			$.each(frm.doc.items || [], function(i, item) {
				frappe.model.set_value(frm.doctype + " Item", item.name, "target_warehouse", frm.doc.target_warehouse);
			});
        }    
    }
});
